src/openforms/appointments/contrib/qmatic/plugin.py

```diff
@@ -40,6 +44,36 @@
 TIMEZONE_AMS = pytz.timezone("Europe/Amsterdam")
 
 
+@contextmanager
+def log_api_errors(template: str, *args):
+    try:
+        yield
+    except (QmaticException, RequestException) as e:
+        logger.exception(template, *args, exc_info=e)
+        raise GracefulQmaticException("SOAP call failed") from e
+    except Exception as exc:
+        raise QmaticException from exc
+
+
+Param = ParamSpec("Param")
+T = TypeVar("T")
+FuncT = Callable[Param, T]
+
+
+def with_graceful_default(default: T):
+    def decorator(func: FuncT) -> FuncT:
+        @wraps(func)
+        def wrapper(*args, **kwargs) -> T:
+            try:
+                return func(*args, **kwargs)
+            except GracefulQmaticException:
+                return default
+
+        return wrapper
+
+    return decorator
```

Comment:
Nice. I was thinking about something like this when I heard about algebraic effects in Unison. And how they use them to do recoverable error handling by throwing/raising with passing a continuation: https://www.unison-lang.org/learn/fundamentals/abilities/error-handling/

Reply:
Took me a bit of frustration when I first tried with a context manager, but that can't affect the return value of the wrapped block, and using a decorator by itself doesn't allow capturing any args you want in the log message (and hence my request to add `structlog` to OF). It only dawned later on me that I can combine them to achieve cleaner error handling.